src/Microsoft.Health.Api/Features/HealthChecks/CachedHealthCheck.cs

```diff
@@ -94,58 +106,78 @@ private async Task<HealthCheckResult> RefreshCache(HealthCheckContext context, D
                 catch (OperationCanceledException oce)
                 {
                     // Re-evaluate the current time and return the cached value if it's still valid
-                    if (!IsCacheValid(Clock.UtcNow))
+                    current = GetCacheSnapshot();
+                    if (!current.IsValid)
                     {
                         _logger.LogWarning(oce, $"Cancellation was requested for {nameof(CheckHealthAsync)}.");
                         throw;
                     }
 
-                    return _cachedResult.Value;
+                    return current.Result;
                 }
                 catch (Exception ex)
                 {
                     _logger.LogError(ex, "Health check failed to complete.");
                     result = HealthCheckResult.Unhealthy(Resources.FailedHealthCheckMessage); // Do not pass error to caller
                 }
 
-                return UpdateCache(result).Value;
+                // Update cache based on the latest snapshot
+                return RefreshCache(GetCacheSnapshot(), result);
             }
             finally
             {
                 _semaphore.Release();
             }
         }
 
-        private CachedHealthCheckResult UpdateCache(HealthCheckResult result)
+        private HealthCheckResult RefreshCache(CacheSnapshot current, HealthCheckResult newResult)
         {
-            // Only update the cache if:
-            // (1) The current result has expired OR
-            // (2) The current result is stale and the cache has been configured to save failed results OR
-            // (3) The current result is stale and the new result is healthy
-            DateTimeOffset currentTime = Clock.UtcNow;
-            if (currentTime >= _cachedResult.ExpireTime || _options.CacheFailure || result.Status != HealthStatus.Unhealthy)
+            // The cache is only updated if appropriate
+            if (current.CanUpdate(newResult.Status, _options.CacheFailure))
             {
-                _cachedResult = new CachedHealthCheckResult
+                var newCache = new CachedHealthCheckResult
                 {
-                    ExpireTime = currentTime + _options.Expiry,
-                    StaleTime = currentTime + _options.Expiry - _options.RefreshOffset,
-                    Value = result,
+                    ExpireTime = current.Timestamp + _options.Expiry,
```

Comment:
What is the difference between Expired vs Stale? Do we need this complication?

Reply:
As discussed offline, it may be helpful to attempt to refresh before the cache fully expires